pkg/services/auth_jwt/validation.go

```diff
@@ -0,0 +1,137 @@
+package auth_jwt
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/grafana/grafana/pkg/models"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+func (s *JWTAuthService) initClaimExpectations() error {
+	if err := json.Unmarshal([]byte(s.Cfg.JWTAuthExpectClaims), &s.expect); err != nil {
+		return err
+	}
+
+	for key, value := range s.expect {
+		switch key {
+		case "iss":
+			if stringValue, ok := value.(string); ok {
+				s.expectRegistered.Issuer = stringValue
+			} else {
+				return fmt.Errorf("%q expectation has invalid type %T, string expected", key, value)
+			}
+			delete(s.expect, key)
+		case "sub":
+			if stringValue, ok := value.(string); ok {
+				s.expectRegistered.Subject = stringValue
+			} else {
+				return fmt.Errorf("%q expectation has invalid type %T, string expected", key, value)
+			}
+			delete(s.expect, key)
+		case "aud":
+			switch value := value.(type) {
+			case []interface{}:
+				for _, val := range value {
+					if val, ok := val.(string); ok {
+						s.expectRegistered.Audience = append(s.expectRegistered.Audience, val)
+					} else {
+						return fmt.Errorf("%q expectation contains value with invalid type %T, string expected", key, val)
+					}
+				}
+			case string:
+				s.expectRegistered.Audience = []string{value}
+			default:
+				return fmt.Errorf("%q expectation has invalid type %T, array or string expected", key, value)
+			}
+			delete(s.expect, key)
+		}
+	}
+
+	return nil
+}
+
+func (s *JWTAuthService) validateClaims(claims models.JWTClaims) error {
+	var registeredClaims jwt.Claims
```

Comment:
Can we also add a test and check to make sure "alg: none" does not work?
https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/